examples/usercall-extension/runner/src/main.rs

```diff
@@ -0,0 +1,116 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extern crate aesm_client;
+extern crate enclave_runner;
+extern crate sgxs_loaders;
+
+use aesm_client::AesmClient;
+use enclave_runner::EnclaveBuilder;
+use enclave_runner::usercalls::{UsercallExtension, SyncStream};
+use sgxs_loaders::isgx::Device as IsgxDevice;
+
+use std::io::{Result as IoResult, Write, Read};
+use std::process::{Command, Child, Stdio};
+use std::sync::{Arc, Mutex};
+
+/// This example demonstrates use of user-call extensions.
+/// User call extension allow the enclave code to "connect" to an external service via a customized enclave runner.
+/// Here we customize the runner to intercept calls to connect to an address "cat" which actually connects the enclave application to 
+/// stdin and stdout of `cat` process.
+
+struct CatService {
+   c : Arc<Mutex<Child>>
+}
+impl CatService {
+   fn new() -> Result<CatService, std::io::Error> {
+        Command::new("/bin/cat")
+            .stdout(Stdio::piped())
+            .stdin(Stdio::piped())
+            .spawn()
+            .map(|c| Arc::new(Mutex::new(c)))
+            .map(|c| CatService {c})
+   }
+}
+impl SyncStream for CatService {
+
+    fn read(&self, buf: &mut [u8]) -> IoResult<usize> {
+        self.c.lock().unwrap().stdout.as_mut().unwrap().read(buf)
+    }
+
+    fn write(&self, buf: &[u8]) -> IoResult<usize> {
+        self.c.lock().unwrap().stdin.as_mut().unwrap().write(buf)
+    }
+
+    fn flush(&self) -> IoResult<()> {
+        self.c.lock().unwrap().stdin.as_mut().unwrap().flush()
+    }
+}
+
+#[derive(Debug)]
+struct ExternalService;
+impl UsercallExtension for ExternalService {
+    fn connect_stream(
+        &self,
+        addr: &[u8],
+        local_addr: Option<&mut String>,
+        peer_addr: Option<&mut String>,
+    ) -> IoResult<Option<Box<SyncStream>>> {
+        let name = String::from_utf8(addr.to_vec()).unwrap_or_else(|_| "def".to_owned());
```

Comment:
Could we change the `UsercallExtension` trait so that `connect_stream` take `addr: &str` rather than `addr: &[u8]`. This way, each implementer won't have to use `String::from_utf8` and handle the error. Also, the default implementation expects `addr` to be a string, so I think it makes sense to enforce this for extensions as well.

Reply:
Thanks. Incorporated in current push.